// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: protobuf/struct.proto

package protobuf

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on WebsocketProto with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *WebsocketProto) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on WebsocketProto with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in WebsocketProtoMultiError,
// or nil if none found.
func (m *WebsocketProto) ValidateAll() error {
	return m.validate(true)
}

func (m *WebsocketProto) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for EventId

	// no validation rules for Payload

	if len(errors) > 0 {
		return WebsocketProtoMultiError(errors)
	}

	return nil
}

// WebsocketProtoMultiError is an error wrapping multiple validation errors
// returned by WebsocketProto.ValidateAll() if the designated constraints
// aren't met.
type WebsocketProtoMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m WebsocketProtoMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m WebsocketProtoMultiError) AllErrors() []error { return m }

// WebsocketProtoValidationError is the validation error returned by
// WebsocketProto.Validate if the designated constraints aren't met.
type WebsocketProtoValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e WebsocketProtoValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e WebsocketProtoValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e WebsocketProtoValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e WebsocketProtoValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e WebsocketProtoValidationError) ErrorName() string { return "WebsocketProtoValidationError" }

// Error satisfies the builtin error interface
func (e WebsocketProtoValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sWebsocketProto.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = WebsocketProtoValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = WebsocketProtoValidationError{}

// Validate checks the field values on WebsocketMessage with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *WebsocketMessage) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on WebsocketMessage with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// WebsocketMessageMultiError, or nil if none found.
func (m *WebsocketMessage) ValidateAll() error {
	return m.validate(true)
}

func (m *WebsocketMessage) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Message

	// no validation rules for Sender

	// no validation rules for Timestamp

	if len(errors) > 0 {
		return WebsocketMessageMultiError(errors)
	}

	return nil
}

// WebsocketMessageMultiError is an error wrapping multiple validation errors
// returned by WebsocketMessage.ValidateAll() if the designated constraints
// aren't met.
type WebsocketMessageMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m WebsocketMessageMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m WebsocketMessageMultiError) AllErrors() []error { return m }

// WebsocketMessageValidationError is the validation error returned by
// WebsocketMessage.Validate if the designated constraints aren't met.
type WebsocketMessageValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e WebsocketMessageValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e WebsocketMessageValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e WebsocketMessageValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e WebsocketMessageValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e WebsocketMessageValidationError) ErrorName() string { return "WebsocketMessageValidationError" }

// Error satisfies the builtin error interface
func (e WebsocketMessageValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sWebsocketMessage.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = WebsocketMessageValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = WebsocketMessageValidationError{}
